Implement MSAA for 2D

Geometror opened the pull request:
This PR implements MSAA for the 2D/Canvas renderer. (Implements godotengine/godot-proposals#1297) In addition, this might improve the situation for godotengine/godot-proposals#1126.

Detailed changes:
- Implement 2D MSAA support for the Canvas renderer, which can be controlled separately from 3D MSAA for maximum flexibility (2X, 4X and 8X are supported, same as for 3D)
	- implementation should be efficient since we can use automatic resolving without the need of VkCmdResolveImage, which has a negative impact on bandwidth
	- can be set for each viewport separately
	- currently not implemented for OpenGL implementation (but should be easy once the OpenGL renderer is done)
- Rename project settings, viewport properties and methods from msaa to msaa_2/3d for clarity
- Minor code restructuring and comment style fixes

![Screenshot 2022-07-29 011539](https://user-images.githubusercontent.com/50084500/181653640-24b89575-8368-49fc-b168-3562d0ce5693.png)

Without MSAA:
![godot windows tools 64_ev8S4G9ANk](https://user-images.githubusercontent.com/50084500/179032809-cfbffb3f-ef88-4cb2-89d7-172d119ab939.png)
MSAA 2x:
![HTVAzzwNeh](https://user-images.githubusercontent.com/50084500/179032904-eb59422a-f651-4be4-af9a-f1e2429f230f.png)
MSAA 4x:
![P7b5duDhEK](https://user-images.githubusercontent.com/50084500/179032919-a0b3fc9e-18d7-4e1f-bf36-644c5f8530ae.png)
MSAA 8x:
![DY33rv7W7a](https://user-images.githubusercontent.com/50084500/179032950-80479d1d-2081-4676-b136-a8b80d8561dd.png)

*Bugsquad edit: This partially addresses https://github.com/godotengine/godot-proposals/issues/4354. This closes https://github.com/godotengine/godot/issues/56233 by adding MSAA sample checks to both 2D and 3D.*

MarianoGnu commented on drivers/gles3/storage/texture_storage.cpp:
```diff
@@ -1553,6 +1553,14 @@ void TextureStorage::render_target_clear_used(RID p_render_target) {
+	RenderTarget *rt = render_target_owner.get_or_null(p_render_target);
+	ERR_FAIL_COND(!rt);
+	rt->msaa = p_msaa;
+	// TODO: Currently not implemented
```
is this TODO real or just a leftover?

Geometror replied:
It's intentional since my initial implementation was only for the Vulkan renderer. But I am currently working on the OpenGL implementation, so I will remove it after I am done :)

MarianoGnu commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -2102,6 +2114,17 @@ void TextureStorage::_update_render_target(RenderTarget *rt) {
+	if (rt->msaa != RS::VIEWPORT_MSAA_DISABLED) {
+		RD::TextureView rd_view_resolve;
+		// Reuse the texture format of the color attachment.
+		rd_format.samples = RD::TEXTURE_SAMPLES_1;
```
i am not understanding, why are you overriding the number of samples from L2106? Given this looks intentional, can oyu be a little more clear on the comment in the previous line?

Geometror replied:
I reuse rd_format since it only functions as a configuration object. The color_resolve buffer should have the same color format/size/etc. but it must have a sample count of 1 since the multisampled color buffer is resolved later into color_resolve. Adjusted the comment.

MarianoGnu commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -2259,6 +2286,13 @@ void TextureStorage::render_target_set_as_unused(RID p_render_target) {
 
+void TextureStorage::render_target_set_msaa(RID p_render_target, RS::ViewportMSAA p_msaa) {
+	RenderTarget *rt = render_target_owner.get_or_null(p_render_target);
+	ERR_FAIL_COND(!rt);
```
can we make an early return here?
```
if (rt->msaa == p_msaa) {
	return;
}
```

Geometror replied:
You're right, we could, but I wasn't done for all the other render_target_set_xxx methods so I'd rather do it in a refactoring PR later.

clayjohn replied:
This one needs the early return because it calls a function. The others don't call functions so they don't need early returns.

MarianoGnu commented on drivers/gles3/storage/texture_storage.cpp:
```diff
@@ -1553,6 +1553,14 @@ void TextureStorage::render_target_clear_used(RID p_render_target) {
 
+void TextureStorage::render_target_set_msaa(RID p_render_target, RS::ViewportMSAA p_msaa) {
+	RenderTarget *rt = render_target_owner.get_or_null(p_render_target);
+	ERR_FAIL_COND(!rt);
```
can we make an early return here?
```
if (rt->msaa == p_msaa) {
	return;
}
```

Geometror commented on editor/editor_node.cpp:
```diff
@@ -541,6 +541,9 @@ void EditorNode::_update_from_settings() {
 	scene_root->set_sdf_scale(sdf_scale);
 
+	Viewport::MSAA msaa = Viewport::MSAA(int(GLOBAL_GET("rendering/anti_aliasing/quality/msaa_2d")));
+	scene_root->set_msaa_2d(msaa);
```
MSAA for 3D is set in `node_3d_editor_plugin`, so it also works in the editor. Unfortunately, in the `canvas_item_editor_plugin `there is currently no callback like `_project_settings_changed`, so I had to place it in `editor_node`

Calinou replied:
3D MSAA is already set in node_3d_editor_plugin.cpp.

Calinou commented on editor/plugins/material_editor_plugin.cpp:
```diff
@@ -151,7 +151,7 @@ MaterialEditor::MaterialEditor() {
 	viewport->set_disable_input(true);
 	viewport->set_transparent_background(true);
-	viewport->set_msaa(Viewport::MSAA_4X);
+	viewport->set_msaa_3d(Viewport::MSAA_4X);
```
I think editor previews should always use 4× MSAA, which is a good compromise between performance and quality. 8× MSAA barely has any visual impact, but it sure has a huge performance impact in modern rendering engines.

Geometror commented on servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:
```diff
@@ -2460,15 +2464,18 @@ RendererCanvasRenderRD::RendererCanvasRenderRD() {
 			Vector<RD::AttachmentFormat> attachments;
 
 			RD::AttachmentFormat af_color;
+			//af_color.samples = RD::TEXTURE_SAMPLES_8;
```
I just noticed that I completely forgot to clean up all changes in `renderer_canvas_render_rd.cpp`.

Geometror commented:
Cleaned it up and added the implementation for the OpenGL renderer:
![grafik](https://user-images.githubusercontent.com/50084500/179362856-b368a14b-d811-40ee-89a8-52fe1ed6cbee.png)

Geometror commented:
Rebased and added checks for the maximum supported sample count for both 2D and 3D MSAA. This should fix any crashes on Mali when MSAA is above 4x.

clayjohn commented on drivers/gles3/storage/texture_storage.cpp:
```diff
@@ -1553,6 +1604,15 @@ void TextureStorage::render_target_clear_used(RID p_render_target) {
+	RenderTarget *rt = render_target_owner.get_or_null(p_render_target);
+	ERR_FAIL_COND(!rt);
+#ifndef JAVASCRIPT_ENABLED
+	rt->msaa = p_msaa;
```
We should likely have an error or warning if the user sets the 2D MSAA differently than the 3D MSAA as they share a framebuffer. 

Once we add Post effects that require a separate pass, we could remove the requirement for the MSAA to be the same and then properly resolve and copy the 3D color texture to the 2D Framebuffer. For now implementing that will be overkill.

clayjohn commented on doc/classes/ProjectSettings.xml:
```diff
@@ -1628,8 +1628,11 @@
-		<member name="rendering/anti_aliasing/quality/msaa" type="int" setter="" getter="" default="0">
-			Sets the number of MSAA samples to use (as a power of two). MSAA is used to reduce aliasing around the edges of polygons. A higher MSAA value results in smoother edges but can be significantly slower on some hardware. See also bilinear scaling 3d [member rendering/scaling_3d/mode] for supersampling, which provides higher quality but is much more expensive.
+		<member name="rendering/anti_aliasing/quality/msaa_2d" type="int" setter="" getter="" default="0">
+			Sets the number of MSAA samples to use for 2D/Canvas rendering (as a power of two). MSAA is used to reduce aliasing around the edges of polygons. A higher MSAA value results in smoother edges but can be significantly slower on some hardware.
```
We should put together a list of what objects users can expect to see AA on and which they won't. For example, users shouldn't expect AA either within or on the edges of sprites. But they may see AA on the edges of sprites where the Texture us solid and goes right to the edge of the quad.

Calinou replied:
Shouldn't MSAA work on the edges of sprites (at least if they're rotated)? Sprites are rects after all, and they're probably drawn with 2 triangles behind the scenes.

Geometror replied:
Added a small note to all MSAA related properties: `This has no effect on shader-induced aliasing or texture aliasing.`

clayjohn reviewed (changes requested):
Great work! I am very excited about this PR and I know many users will be too. 

I have left a few comments on implementation, mostly just stuff that needs to be fixed up in the GLES3 implementation.

Geometror reviewed (commented):
Currently combined 2D and 3D rendering is broken since the resolve attachment's store op is `VK_ATTACHMENT_STORE_OP_DONT_CARE` while it should be `VK_ATTACHMENT_STORE_OP_STORE`. (this took way too long to figure out :)) 
![grafik](https://user-images.githubusercontent.com/50084500/181396214-0f0a0ad1-e133-4d59-ab2c-bb657def724e.png)
Just brute forcing all store ops to `VK_ATTACHMENT_STORE_OP_STORE` resolves the issue:
![grafik](https://user-images.githubusercontent.com/50084500/181395672-4c1893d6-2284-46ef-b13c-7e8eec448bed.png)
Now that I know what's wrong I can work on a proper fix, will probably push one tomorrow after I addressed clayjohns review comments. (I also rewrote a few bits to make things cleaner)

Geometror commented:
The Vulkan side is should be done now, OpenGL still needs some work, but because of #63600 I can't test my changes - so that might take some time. If it's ok, I would prefer to only have the critical parts in the OpenGL implementation fixed and then implement 2D MSAA properly later (e.g. with webgl2 support) when the OpenGL renderer has matured. This PR was already pretty time-consuming and unfortunately I have not much time currently.
![Screenshot 2022-07-29 011539](https://user-images.githubusercontent.com/50084500/181653640-24b89575-8368-49fc-b168-3562d0ce5693.png)

Geometror commented:
Rebased. Regarding the GLES3 part: I just looked at the code and noticed that 3D MSAA is also not yet implemented. I would rather strip the implementation completely from this PR and add both 2D and 3D MSAA together at a later point in time (and properly, so that you can set both independently from each other like in the Vulkan renderer).

clayjohn commented:
> Rebased. Regarding the GLES3 part: I just looked at the code and noticed that 3D MSAA is also not yet implemented. I would rather strip the implementation completely from this PR and add both 2D and 3D MSAA together at a later point in time (and properly, so that you can set both independently from each other like in the Vulkan renderer).

Fair enough. I am okay with that if you are planning on adding it back at a later time to avoid this getting too complex.

Geometror commented:
Ok, I removed the GLES3 implementation (a warning is printed when the user tries to enable it). With that I think this PR is complete :)

clayjohn commented on drivers/vulkan/rendering_device_vulkan.cpp:
```diff
@@ -8900,6 +8908,46 @@ void RenderingDeviceVulkan::_begin_frame() {
 
+VkSampleCountFlagBits RenderingDeviceVulkan::_ensure_supported_sample_count(TextureSamples p_requested_sample_count) const {
+	VkSampleCountFlags sample_count_flags = limits.framebufferColorSampleCounts & limits.framebufferDepthSampleCounts;
+	VkSampleCountFlagBits max_sample_count = VK_SAMPLE_COUNT_1_BIT;
+
+	if (sample_count_flags & VK_SAMPLE_COUNT_2_BIT) {
+		max_sample_count = VK_SAMPLE_COUNT_2_BIT;
+	}
+	if (sample_count_flags & VK_SAMPLE_COUNT_4_BIT) {
+		max_sample_count = VK_SAMPLE_COUNT_4_BIT;
+	}
+	if (sample_count_flags & VK_SAMPLE_COUNT_8_BIT) {
+		max_sample_count = VK_SAMPLE_COUNT_8_BIT;
+	}
+	if (sample_count_flags & VK_SAMPLE_COUNT_16_BIT) {
+		max_sample_count = VK_SAMPLE_COUNT_16_BIT;
+	}
+	if (sample_count_flags & VK_SAMPLE_COUNT_32_BIT) {
+		max_sample_count = VK_SAMPLE_COUNT_32_BIT;
+	}
+	if (sample_count_flags & VK_SAMPLE_COUNT_64_BIT) {
+		max_sample_count = VK_SAMPLE_COUNT_64_BIT;
+	}
+
+	return MIN(max_sample_count, rasterization_sample_count[p_requested_sample_count]);
```
This should all be deleted now as it is right now the new code will never be reached

Geometror replied:
Oops, I don't know how I overlooked that. Fixed :)

clayjohn reviewed (changes requested):
Just the one change before it is ready to go

clayjohn reviewed (approved):
I'm happy with this PR. Should be ready to merge. 

CC @BastiaanOlij I think this may conflict with your Framebuffer reorganization PR, but I think we should merge it now as Geometror won't be available for awhile after the end of this week.

Geometror committed:
Implement MSAA for 2D [Vulkan only]

reduz commented:
Does this work properly with backbuffer or clip effects? I think it would be good to test this before merging.

Geometror commented:
Yes, this seems to work fine with those (MSAA 8x):

Polygon2D with clip children enabled
![image](https://user-images.githubusercontent.com/50084500/185094734-e30c8036-466d-4bb2-aa24-0ca42f2388d4.png)

Backbuffer-reading post processing shader
![image](https://user-images.githubusercontent.com/50084500/185094307-bc898572-0ca8-4d89-b0de-0c25ec8b384f.png)

akien-mga commented:
@reduz says it's OK so let's merge :+1:

akien-mga commented:
Thanks!

insomniacUNDERSCORElemon commented:
Small note: this doesn't smooth the clipped items themselves, only the parent. Note the zoomed square:

![ArcoLinux_2022-09-17_04-38-17](https://user-images.githubusercontent.com/4397725/190848306-434f7e67-53ef-4a08-82bc-22ecafca96be.png)

(also seen w/this screenshot is the backbuffer ghosting which is not specific to MSAA, also unwanted color blending)

Calinou commented:
I wonder if 2D clipping is performed in a shader. If so, would alpha-to-coverage make it able to be antialiased?

Geometror commented:
@insomniacUNDERSCORElemon Could you open an issue and provide an reproduction project? :)

Calinou commented:
> @insomniacUNDERSCORElemon Could you open an issue and provide an reproduction project? :)

https://github.com/godotengine/godot/issues/66005
